### ISSUE TITLE
`AndroidNativeLibrary` doesn't trigged `JNI_OnLoad`

### ISSUE THREAD

**Issue by thenextman** [dotnet/android#7616]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

VS 2022 17.4, Xamarin.Android 13.1.0.1

### Description

I have a shared library (.so) written in C. The library exports `JNI_OnLoad` and `JNI_OnUnload`, confirmed with `readelf`:

>  478: 00000000002ff2f0   132 FUNC    GLOBAL DEFAULT   13 JNI_OnLoad
>  3111: 00000000002ff380   297 FUNC    GLOBAL DEFAULT   13 JNI_OnUnload

I reference the .so properly in my Xamarin.Android application using `AndroidNativeLibrary`.

However, at runtime `JNI_OnLoad` is never invoked.

### Steps to Reproduce

See above

### Did you find any workaround?

I _can_ invoke `JNI_OnLoad` manually by first using reflection to get the `java_vm` pointer:

```
[DllImport("MyLibrary", CallingConvention = CallingConvention.Cdecl)]
private static extern int JNI_OnLoad(IntPtr vm, IntPtr reserved);

static MyClass()
{
    var field = typeof(Android.Runtime.JNIEnv).GetField("java_vm", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static);
    JNI_OnLoad((IntPtr)field.GetValue(null), IntPtr.Zero);
}
```

But this feels like an ugly workaround.

### Relevant log output

```shell
N/A
```

**Comment by grendello** [comment]
@thenextman the library needs to be loaded explicitly, most likely, in the similar way [we load](https://github.com/xamarin/xamarin-android/blob/b165bc86c0026b17d0aacb1c1fd81afe9a05c762/src/java-runtime/java/mono/android/MonoPackageManager.java#L113) Xamarin.Android's runtime (`libmonodroid.so`).

**Comment by thenextman** [comment]
@grendello Thanks for the input.

I do see my library (libMyLib.so) packaged inside {app.apk}/lib/{abi} alongside `libmonodroid` et all, as in your example. It exists for all supported ABIs.

However calling `Java.Lang.JavaSystem.LoadLibrary("MyLib");` yields:

> Java.Lang.UnsatisfiedLinkError: dlopen failed: library "libMyLib.so" not found
>   at Java.Interop.JniEnvironment+StaticMethods.CallStaticVoidMethod (Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00068] in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/JniEnvironment.g.cs:13250 
>   at Java.Interop.JniPeerMembers+JniStaticMethods.InvokeVoidMethod (System.String encodedMember, Java.Interop.JniArgumentValue* parameters) [0x00008] in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniStaticMethods.cs:97 
>   at Java.Lang.JavaSystem.LoadLibrary (System.String libname) [0x0001f] in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-31/mcw/Java.Lang.JavaSystem.cs:339 

I *can* p/invoke the library (`[DllImport("MyLib")]...`) but equally, `JNI_OnLoad` is not invoked.

> the first time Java resolves a native function to be in one of the loaded libraries

I'm not sure I follow (I don't know Xamarin.Android internals at all). Where is Java involved here if I'm p/invoking a vanilla C function from .NET?

**Comment by grendello** [comment]
@thenextman let's start with the logcat :) Can you please run your app until it throws the above exception, after first configuring your device as follows:

```shell
$ adb shell setprop debug.mono.log default,assembly,timing=bare
$ adb logcat -G 16M
$ adb logcat -c
# Run your app here, make it crash, wait a few moments and then
$ adb logcat -d > log.txt
```

This will show is why the library fails to load (the "not found" message might be a red herring).

With regards to your question where Java is involved. `JNI_OnLoad` is a Java Native Interface (JNI) function called by the Java VM whenever it encounters a Java class with a function declared as follows:

```java
package helloandroid;


public class MainActivity
        extends androidx.appcompat.app.AppCompatActivity
        implements
                mono.android.IGCUserPeer
{
        public void onCreate (android.os.Bundle p0)
        {
                n_onCreate (p0);
        }

        private native void n_onCreate (android.os.Bundle p0);
}
```

The `native` keyword will cause the Java runtime (ART in Android) to try to find a native shared library which exports a symbol with a properly formatted name. In case of the native `n_onCreate` method, the native symbol name is going to be `Java_helloandroid_MainActivity_n_1onCreate`. Now, for each managed class which eventually derives from the `Java.Lang.Object` Xamarin.Android will generate a corresponding Java class (the above fragment mirrors a managed class `helloandroid.MainActivity`) and that class is what Java VM is going to interact with - every Managed overridden method (or property) will have a corresponding `native` Java method, such as `n_onCreate` above. You can find all the generated Java sources in the `obj` directory of your app, if you're curious how they look.

When the `native` method is called (in our example it is the Xamarin.Android app main activity's `OnCreate` method), Java searches for an exported symbol with that name in all the loaded shared libraries and, when found, determines whether or not it has already called any functions in the DSO containing the symbol. If the answer is "no", then the library is initialized - among others, `JNI_OnLOad` is called and then, eventually, the `native` method is invoked (in Xamarin.Android this will cause control to be transferred to the managed land, eventually).

**Comment by thenextman** [comment]
Hello @grendello. Thank you again for the troubleshooting and detailed info.

With `logcat`, it's strange because I __do__ see detailed assembly loading and binding output for both managed and unmanaged libraries in the .apk. _However_ MyLib (the only library I'm explicitly doing `LoadLibrary`), it gives no extra information:

> 12-06 02:50:30.971 28338 28379 D monodroid: typemap: Java type 'android/net/ConnectivityManager' corresponds to managed type 'Android.Net.ConnectivityManager, Mono.Android'
> 12-06 02:50:30.971 28338 28379 I monodroid-timing: [1/5] Typemap.java_to_managed: end, total time; elapsed: 0:0::66167
> 12-06 02:50:30.973 28338 28379 D monodroid: typemap: Java type 'android/net/NetworkInfo' corresponds to managed type 'Android.Net.NetworkInfo, Mono.Android'
> 12-06 02:50:30.973 28338 28379 I monodroid-timing: [1/5] Typemap.java_to_managed: end, total time; elapsed: 0:0::34708
> 12-06 02:50:30.988 28338 28379 D monodroid: typemap: Java type 'android/hardware/display/DisplayManager' corresponds to managed type 'Android.Hardware.Display.DisplayManager, Mono.Android'
> 12-06 02:50:30.988 28338 28379 I monodroid-timing: [1/5] Typemap.java_to_managed: end, total time; elapsed: 0:0::52667
> 12-06 02:50:31.257 28338 28379 I monodroid-timing: [1/6] Typemap.managed_to_java: end, total time; elapsed: 0:0::3459
> 12-06 02:50:31.258 28338 28379 I monodroid-timing: [1/6] Typemap.managed_to_java: end, total time; elapsed: 0:0::1083
> 12-06 02:50:31.258 28338 28379 I monodroid-timing: [1/10] Runtime.register: end time. Registered type: MyClass, MyApp; elapsed: 0:0::53917
> 12-06 02:50:31.264 28338 28379 D monodroid: typemap: Java type 'java/lang/UnsatisfiedLinkError' corresponds to managed type 'Java.Lang.UnsatisfiedLinkError, Mono.Android'
> 12-06 02:50:31.264 28338 28379 I monodroid-timing: [1/5] Typemap.java_to_managed: end, total time; elapsed: 0:0::68375
> 12-06 02:50:31.265 28338 28372 D OpenGLRenderer: endAllActiveAnimators on 0xb400007bb198b3a0 (MenuPopupWindow$MenuDropDownListView) with handle 0xb400007ac32d96b0
> 12-06 02:50:31.265  1226  1860 W InputManager-JNI: Input channel object 'a02716d PopupWindow:784f6e (client)' was disposed without first being removed with the input manager!
> 12-06 02:50:31.268  2462  2462 I GoogleInputMethodService: GoogleInputMethodService.onFinishInput():3220 
> 12-06 02:50:31.268  2462  2462 I GoogleInputMethodService: GoogleInputMethodService.updateDeviceLockedStatus():2085 repeatCheckTimes = 0, unlocked = true
> 12-06 02:50:31.269  2462  2462 I GoogleInputMethodService: GoogleInputMethodService.onStartInput():1876 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=MyApp fieldId=-1 fieldName=null extras=null}, false)
> 12-06 02:50:31.269  2462  2462 I GoogleInputMethodService: GoogleInputMethodService.shouldHideHeaderOnInitialState():3985 ShouldHideHeaderOnInitialState = false
> 12-06 02:50:31.269  2462  2462 I GoogleInputMethodService: GoogleInputMethodService.updateDeviceLockedStatus():2085 repeatCheckTimes = 2, unlocked = true
> 12-06 02:50:31.272 28338 28379 V mono-stdout: Java.Lang.UnsatisfiedLinkError: dlopen failed: library "libMyLib.so" not found
> 12-06 02:50:31.272 28338 28379 V mono-stdout:   at Java.Interop.JniEnvironment+StaticMethods.CallStaticVoidMethod (Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00068] in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/JniEnvironment.g.cs:13250 
> 12-06 02:50:31.272 28338 28379 V mono-stdout:   at Java.Interop.JniPeerMembers+JniStaticMethods.InvokeVoidMethod (System.String encodedMember, Java.Interop.JniArgumentValue* parameters) [0x00008] in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniStaticMethods.cs:97 
> 12-06 02:50:31.272 28338 28379 V mono-stdout:   at Java.Lang.JavaSystem.LoadLibrary (System.String libname) [0x0001f] in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-31/mcw/Java.Lang.JavaSystem.cs:339 
> 12-06 02:50:31.272 28338 28379 V mono-stdout:   at [my class constructor]
> 12-06 02:50:31.272 28338 28379 V mono-stdout:   --- End of managed Java.Lang.UnsatisfiedLinkError stack trace ---
> 12-06 02:50:31.272 28338 28379 V mono-stdout: java.lang.UnsatisfiedLinkError: dlopen failed: library "libMyLib.so" not found
> 12-06 02:50:31.272 28338 28379 V mono-stdout: 	at java.lang.Runtime.loadLibrary0(Runtime.java:1088)
> 12-06 02:50:31.272 28338 28379 V mono-stdout: 	at java.lang.Runtime.loadLibrary0(Runtime.java:998)
> 12-06 02:50:31.272 28338 28379 V mono-stdout: 	at java.lang.System.loadLibrary(System.java:1661)
> 12-06 02:50:31.272 28338 28379 V mono-stdout: 
> 12-06 02:50:31.297 28338 28372 D EGL_emulation: app_time_stats: avg=34.36ms min=1.62ms max=1019.01ms count=45

I appreciate the extra explanation for the Java interop case. It's very clear. However, it won't apply in my case since I don't call native functions using JNI (no `native` declarations); rather, my native code wants to call Java. So we might call `AttachCurrentThread`, call some Java system libraries, then `DetachCurrentThread`. Of course this requires access to the JNI environment - hence my post here.

 You originally wrote

> just p/invoking into the library should work to trigger JNI_OnLoad

But to be clear, you're talking about `native` decorated functions, right? Since if I just make a p/invoke directly to C without involving Java, JNI isn't involved and it won't do any automatic bootstrapping of my library. Or is there some extra magic behind `DllImport`?

**Comment by grendello** [comment]
> With `logcat`, it's strange because I **do** see detailed assembly loading and binding output for both managed and unmanaged libraries in the .apk. _However_ MyLib (the only library I'm explicitly doing `LoadLibrary`), it gives no extra information:

I hoped for no output for the call to load your library with `LoadLibrary` :) It means it was loaded cleanly (no unresolved symbols etc).  For it to show up in one of our messages, you'd need to load it via Mono - by p/invoking some function in the library, then it would get loaded by our runtime instead of the Java shared library loader.

> 
> I appreciate the extra explanation for the Java interop case. It's very clear. However, it won't apply in my case since I don't call native functions using JNI (no `native` declarations); rather, my native code wants to call Java. So we might call `AttachCurrentThread`, call some Java system libraries, then `DetachCurrentThread`. Of course this requires access to the JNI environment - hence my post here.
> 
OK, so if your DSO doesn't contain any `native` function implementations, then `JNI_OnLoad` won't ever be called, since Java doesn't care about the library. Off the top of my head I don't see a way to obtain `JavaVM*`, but I'll try to figure something out and update the issue once I have something.

> You originally wrote
> 
> > just p/invoking into the library should work to trigger JNI_OnLoad
> 
> But to be clear, you're talking about `native` decorated functions, right? Since if I just make a p/invoke directly to C without involving Java, JNI isn't involved and it won't do any automatic bootstrapping of my library. Or is there some extra magic behind `DllImport`?

Yes, I meant the Java "p/invoke", shouldn't have used this term, sorry for the confusion :)

**Comment by grendello** [comment]
It seems I was wrong - `JNI_OnLoad` **should** be called if you use `LoadLibrary` - the only question then is why `LoadLibrary` cannot find the DSO.

**Comment by grendello** [comment]
@thenextman let's try one other thing. Please issue the following command in addition to the previously mentioned `adb` commands, before starting the app:

```shell
$ adb shell setprop debug.ld.all dlopen,dlerror,dlsym
```

This should enable internal linker logging and we **might** see some errors regarding the library.

**Comment by jonpryor** [comment]
This should *not* be considered an appropriate fix or workaround -- the correct fix is to determine why `JavaSystem.LoadLibrary("MyLib")` is throwing an `UnsatisfiedLinkError`, which is *really* weird -- but there is a public API to get the `JavaVM*` value, via [`JniRuntime.InvocationPointer`](https://github.com/xamarin/java.interop/blob/09f8da26b6124536391153728ed5018465dc9b74/src/Java.Interop/Java.Interop/JniRuntime.cs#L156):

```csharp
IntPtr javaVMPtr = global::Java.Interop.JniRuntime.CurrentRuntime.InvocationPointer;
```

**Comment by thenextman** [comment]
Thanks for the tip @jonpryor 

@grendello Here is the output:

> 12-06 07:46:35.699 29838 29886 D linker  : dlopen(name="libMyLib.so", flags=0x2, extinfo=[flags=0x200, reserved_addr=0x0, reserved_size=0x0, relro_fd=0, library_fd=0, library_fd_offset=0x0, library_namespace=(default)@0x7d65d4e4e0], caller="/apex/com.android.art/lib64/libnativebridge.so", caller_ns=com_android_art@0x7d64a18170, targetSdkVersion=31) ...
> 12-06 07:46:35.700 29838 29886 D linker  : find_libraries(ns=(default)): task=libMyLib.so, is_dt_needed=0
> 12-06 07:46:35.700 29838 29886 D linker  : load_library(ns=(default), task=libMyLib.so, flags=0x2, search_linked_namespaces=1): calling open_library
> 12-06 07:46:35.700 29838 29886 D linker  : library "libMyLib.so" not found
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_internal(ns=(default), task=libMyLib.so): Trying 6 linked namespaces
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_adbd, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_i18n, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_art, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_resolv, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_neuralnetworks, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : find_library_in_linked_namespace(ns=com_android_os_statsd, task=libDevolutionsRdp.so): Not accessible (soname=libMyLib.so)
> 12-06 07:46:35.700 29838 29886 D linker  : ... dlopen failed: library "libMyLib.so" not found
> 12-06 07:46:35.700 29838 29886 D linker  : dlerror set to "dlopen failed: library "libMyLib.so" not found"
>

**Comment by grendello** [comment]
@thenextman if I'm reading this correctly, the load fails because DSO linker cannot find the `libDevolutionsRdp.so` library - is this something your DSO references? You can check with:

```shell
$ llvm-readelf -a YourLibrary.so | grep NEEDED
```

If it is indeed referenced, can you make sure it's contained in the apk alongside your library? What's puzzling, however, is that your attempt to p/invoke into your library was successful, suggesting the `libDevolutionsRdp.so` is indeed there. And yet... Let's check that :)

**Comment by thenextman** [comment]
> @thenextman if I'm reading this correctly, the load fails because DSO linker cannot find the `libDevolutionsRdp.so` library - is this something your DSO references? You can check with:
> 
> 
> 
> ```shell
> 
> $ llvm-readelf -a libMyLib.so | grep NEEDED
> 
> ```
> 
> 
> 
> If it is indeed referenced, can you make sure it's contained in the apk alongside your library? What's puzzling, however, is that your attempt to p/invoke into your library was successful, suggesting the `libDevolutionsRdp.so` is indeed there. And yet... Let's check that :)
> 
> 

@grendello sorry for misleading log. In all cases it's 'libMyLib'; I made a mistake redacting the output. So the "task" and the "so name" are the same.

**Comment by thenextman** [comment]
Hi @grendello 

Lib directory:

> -rw-r--r--  6.3 unx  2747404 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libxamarin-app.so
> -rw-r--r--  6.3 unx  2747032 b- defX 22-Dec-06 14:43 lib/x86/libxamarin-app.so
> -rw-r--r--  6.3 unx  4040160 b- defX 22-Dec-06 14:43 lib/x86_64/libxamarin-app.so
> -rw-r--r--  6.3 unx  3876528 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libxamarin-app.so
> -rwxr-xr-x  6.3 unx   288572 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libmonodroid.so
> -rwxr-xr-x  6.3 unx   336052 b- defX 22-Dec-06 14:43 lib/x86/libmonodroid.so
> -rwxr-xr-x  6.3 unx   338752 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libmonodroid.so
> -rwxr-xr-x  6.3 unx   353712 b- defX 22-Dec-06 14:43 lib/x86_64/libmonodroid.so
> -rwxr-xr-x  6.3 unx    49004 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libxa-internal-api.so
> -rwxr-xr-x  6.3 unx    61056 b- defX 22-Dec-06 14:43 lib/x86/libxa-internal-api.so
> -rwxr-xr-x  6.3 unx    66184 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libxa-internal-api.so
> -rwxr-xr-x  6.3 unx    57608 b- defX 22-Dec-06 14:43 lib/x86_64/libxa-internal-api.so
> -rwxr-xr-x  6.3 unx  1112688 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libmono-btls-shared.so
> -rwxr-xr-x  6.3 unx  1459584 b- defX 22-Dec-06 14:43 lib/x86/libmono-btls-shared.so
> -rwxr-xr-x  6.3 unx  1613872 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libmono-btls-shared.so
> -rwxr-xr-x  6.3 unx  1885864 b- defX 22-Dec-06 14:43 lib/x86_64/libmono-btls-shared.so
> -rwxr-xr-x  6.3 unx   272856 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libmono-profiler-log.so
> -rwxr-xr-x  6.3 unx   261792 b- defX 22-Dec-06 14:43 lib/x86/libmono-profiler-log.so
> -rwxr-xr-x  6.3 unx   284568 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libmono-profiler-log.so
> -rwxr-xr-x  6.3 unx   301928 b- defX 22-Dec-06 14:43 lib/x86_64/libmono-profiler-log.so
> -rwxr-xr-x  6.3 unx  4456548 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libmonosgen-2.0.so
> -rwxr-xr-x  6.3 unx  4212360 b- defX 22-Dec-06 14:43 lib/x86/libmonosgen-2.0.so
> -rwxr-xr-x  6.3 unx  4039176 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libmonosgen-2.0.so
> -rwxr-xr-x  6.3 unx  4673000 b- defX 22-Dec-06 14:43 lib/x86_64/libmonosgen-2.0.so
> -rwxr--r--  6.3 unx  5612008 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libMyLib.so
> -rwxr--r--  6.3 unx  7762984 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libMyLib.so
> -rwxr--r--  6.3 unx  7183136 b- defX 22-Dec-06 14:43 lib/x86/libMyLib.so
> -rwxr--r--  6.3 unx  7895792 b- defX 22-Dec-06 14:43 lib/x86_64/libMyLib.so
> -rwxr-xr-x  6.3 unx   811380 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libsqlite3_xamarin.so
> -rwxr-xr-x  6.3 unx  1270380 b- defX 22-Dec-06 14:43 lib/x86/libsqlite3_xamarin.so
> -rwxr-xr-x  6.3 unx  1172352 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libsqlite3_xamarin.so
> -rwxr-xr-x  6.3 unx  1273144 b- defX 22-Dec-06 14:43 lib/x86_64/libsqlite3_xamarin.so
> -rwxr-xr-x  6.3 unx   785140 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libmono-native.so
> -rwxr-xr-x  6.3 unx   823560 b- defX 22-Dec-06 14:43 lib/x86/libmono-native.so
> -rwxr-xr-x  6.3 unx   750976 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libmono-native.so
> -rwxr-xr-x  6.3 unx   880632 b- defX 22-Dec-06 14:43 lib/x86_64/libmono-native.so
> -rwxr-xr-x  6.3 unx    31672 b- defX 22-Dec-06 14:43 lib/armeabi-v7a/libxamarin-debug-app-helper.so
> -rwxr-xr-x  6.3 unx    31472 b- defX 22-Dec-06 14:43 lib/x86/libxamarin-debug-app-helper.so
> -rwxr-xr-x  6.3 unx    37336 b- defX 22-Dec-06 14:43 lib/arm64-v8a/libxamarin-debug-app-helper.so
> -rwxr-xr-x  6.3 unx    36280 b- defX 22-Dec-06 14:43 lib/x86_64/libxamarin-debug-app-helper.so

I do confirm that the properly library ABIs are in the proper directories (and this is double-confirmed by my ability to p/invoke into the native code using `DllImport`).

Dependencies:

> libMyLib.so | grep NEEDED
>  0x0000000000000001 (NEEDED)             Shared library: [liblog.so]
>  0x0000000000000001 (NEEDED)             Shared library: [libdl.so]
>  0x0000000000000001 (NEEDED)             Shared library: [libm.so]
>  0x0000000000000001 (NEEDED)             Shared library: [libOpenSLES.so]
>  0x0000000000000001 (NEEDED)             Shared library: [libc.so]

We need just a few system libraries, everything else is statically linked into the shared object.

Header:

> ELF Header:
>   Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
>   Class:                             ELF64
>   Data:                              2's complement, little endian
>   Version:                           1 (current)
>   OS/ABI:                            UNIX - System V
>   ABI Version:                       0
>   Type:                              DYN (Shared object file)
>   Machine:                           AArch64
>   Version:                           0x1
>   Entry point address:               0x0
>   Start of program headers:          64 (bytes into file)
>   Start of section headers:          6675576 (bytes into file)
>   Flags:                             0x0
>   Size of this header:               64 (bytes)
>   Size of program headers:           56 (bytes)
>   Number of program headers:         8
>   Size of section headers:           64 (bytes)
>   Number of section headers:         27
>   Section header string table index: 26

Manifest:

> <?xml version="1.0" encoding="utf-8" standalone="no"?><manifest xmlns:android="http://schemas.android.com/apk/res/android" android:compileSdkVersion="31" android:compileSdkVersionCodename="12" android:installLocation="auto" package="com.mycompany.myapp" platformBuildVersionCode="31" platformBuildVersionName="12">
>     <uses-feature android:name="android.hardware.telephony" android:required="false"/>
>     <uses-permission android:name="android.permission.INTERNET"/>
>     <uses-permission android:name="android.permission.READ_LOGS"/>
>     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
>     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
>     <uses-permission android:name="android.permission.INSTALL_PACKAGES"/>
>     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
>     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
>     <uses-permission android:name="android.permission.USE_FINGERPRINT"/>
>     <uses-permission android:name="android.permission.CALL_PHONE"/>
>     <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
>     <uses-permission android:name="android.permission.NFC"/>
>     <uses-permission android:name="android.permission.USE_FACE"/>
>     <uses-permission android:name="android.permission.USE_BIOMETRIC"/>
>     <uses-permission android:name="android.permission.RECORD_AUDIO"/>
>     <queries>
>         <package android:name="com.google.android.youtube"/>
>         <intent>
>             <action android:name="android.intent.action.VIEW"/>
>             <category android:name="android.intent.category.BROWSABLE"/>
>             <data android:scheme="https"/>
>         </intent>
>     </queries>
>     <queries>
>         <intent>
>             <action android:name="android.intent.action.DIAL"/>
>             <category android:name="android.intent.category.DEFAULT"/>
>             <data android:scheme="tel"/>
>         </intent>
>     </queries>
>     <queries>
>         <intent>
>             <action android:name="android.intent.action.SENDTO"/>
>             <data android:scheme="mailto"/>
>             <category android:name="android.intent.category.DEFAULT"/>
>         </intent>
>     </queries>
>     <application android:allowBackup="true" android:appComponentFactory="androidx.core.app.CoreComponentFactory" android:debuggable="true" android:extractNativeLibs="true" android:icon="@drawable/icon" android:label="My App" android:name="androidx.multidex.MultiDexApplication" android:requestLegacyExternalStorage="true" android:theme="@style/MyTheme" android:usesCleartextTraffic="true">
>         <activity android:configChanges="keyboard|orientation" android:exported="false" android:launchMode="singleTask" android:name="com.dropbox.client2.android.AuthActivity">
>             <intent-filter>
>                 <data android:scheme="db-03gcqj1xs9f2n72"/>
>                 <action android:name="android.intent.action.VIEW"/>
>                 <category android:name="android.intent.category.BROWSABLE"/>
>                 <category android:name="android.intent.category.DEFAULT"/>
>             </intent-filter>
>         </activity>
>         <receiver android:name="crc64a274fa7acfcbef02.ProtocolNotificationActionReceiver"/>
>         <provider android:authorities="com.mycompany.myapp.fileprovider" android:exported="false" android:grantUriPermissions="true" android:name="androidx.core.content.FileProvider">
>             <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/attachments_file_paths"/>
>         </provider>
>         <meta-data android:name="android.content.APP_RESTRICTIONS" android:resource="@xml/app_air_watch_settings_management"/>
>         <meta-data android:name="com.samsung.android.multidisplay.keep_process_alive" android:value="true"/>
>         <service android:name="crc64a274fa7acfcbef02.ProtocolBackgroundService"/>
>         <service android:exported="false" android:label="My App" android:name="com.mycompany.myapp.Autofill.AutofillService" android:permission="android.permission.BIND_AUTOFILL_SERVICE">
>             <meta-data android:name="android.autofill" android:resource="@xml/browser_compatibility_list"/>
>             <intent-filter>
>                 <action android:name="android.service.autofill.AutofillService"/>
>             </intent-filter>
>         </service>
>         <activity android:exported="true" android:name="crc647b689e195de094fb.SplashScreenActivity" android:noHistory="false" android:resizeableActivity="true" android:theme="@style/Theme.SplashNew">
>             <intent-filter>
>                 <action android:name="android.intent.action.MAIN"/>
>                 <category android:name="android.intent.category.LAUNCHER"/>
>             </intent-filter>
>         </activity>
>         <activity android:name="crc64bea9cdcd06864287.MySessionActivity"/>
>         <service android:name="crc64396a3fe5f8138e3f.KeepAliveService"/>
>         <provider android:authorities="com.mycompany.myapp.mono.android.MultiDexLoader.__mono_init__" android:exported="false" android:initOrder="1999999999" android:name="mono.android.MultiDexLoader"/>
>         <provider android:authorities="com.mycompany.myapp.mono.MonoRuntimeProvider.__mono_init__" android:exported="false" android:initOrder="1999999998" android:name="mono.MonoRuntimeProvider"/>
>         <provider android:authorities="com.mycompany.myapp.androidx-startup" android:exported="false" android:name="androidx.startup.InitializationProvider">
>             <meta-data android:name="androidx.lifecycle.ProcessLifecycleInitializer" android:value="androidx.startup"/>
>             <meta-data android:name="androidx.emoji2.text.EmojiCompatInitializer" android:value="androidx.startup"/>
>         </provider>
>         <uses-library android:name="androidx.window.extensions" android:required="false"/>
>         <uses-library android:name="androidx.window.sidecar" android:required="false"/>
>     </application>
> </manifest>

**Comment by thenextman** [comment]
@grendello No change either with release build or `extractNativeLibs="false"`.

Unless you have some other ideas, my next step will be to try and strip this down to a minimal repro that I can share here.

**Comment by grendello** [comment]
@thenextman I'm out of ideas, my next step was to ask you for a self-contained repro... :) Thanks!

**Comment by thenextman** [comment]
@grendello I stripped this down to a simple example, and discovered that `LoadLibrary` fails in this manner if called from a background thread. If I marshal the call back to the UI thread, things work as expected.

I didn't see this mentioned in any documentation (Xamarin or Android) - is it a known requirement of calling `LoadLibrary`? From Java or Xamarin? I see no reason I shouldn't be able to use this from a background thread, but I apologize if I missed some subtlety here.

**Comment by thenextman** [comment]
The Google [docs](https://developer.android.com/training/articles/perf-jni#native-libraries) mention that `LoadLibrary` should be called "from a static class initializer" (although they don't tell if this is a hard requirement or just a recommendation - or why). They also don't mention threading as a potential cause of `UnsatisifiedLinkError`.

I have to assume it's safe to call from a thread using Java.

**Comment by grendello** [comment]
@thenextman this is quite a discovery! I wasn't aware about the thread restriction for `LoadLibrary` and as far as XA/Mono are concerned, there's no such requirement (you can p/invoke into or dlopen libraries from any thread). 

The "static constructor" remark makes sense if you consider that the library being loaded might contain implementations for the `native` Java methods in the loading class, but it doesn't apply here IMO.

So, you should be able to load the library from your activity's `OnCreate` without any tricks?

I'd say this find allows us to close this issue?

**Comment by thenextman** [comment]
> So, you should be able to load the library from your activity's OnCreate without any tricks?

Indeed, or anywhere I have the ability to marshal back to the UI thread.

> I'd say this find allows us to close this issue?

I'd say so, yes. Thank you for the help and insights!

**Comment by jonpryor** [comment]
I tested, and pure Java/Kotlin does *not* have the same restriction.  We're missing something.

Start with: https://github.com/android/ndk-samples/tree/eb1203395e9813a21fa27652ded7468dec074d51/hello-jni

Apply this patch:

```diff
diff --git a/hello-jni/app/src/main/cpp/hello-jni.cpp b/hello-jni/app/src/main/cpp/hello-jni.cpp
index 2cd9916..b2aa375 100644
--- a/hello-jni/app/src/main/cpp/hello-jni.cpp
+++ b/hello-jni/app/src/main/cpp/hello-jni.cpp
@@ -15,9 +15,20 @@
  *
  */
 #include <jni.h>
+#include <android/log.h>
 
 #include <string>
 
+static JavaVM *gvm;
+
+extern "C" JNIEXPORT jint
+JNI_OnLoad(JavaVM *vm, void *reserved)
+{
+  __android_log_print (ANDROID_LOG_INFO, "*jonp*", "JNI_OnLoad called!");
+  gvm = vm;
+  return JNI_VERSION_1_6;
+}
+
 extern "C" JNIEXPORT jstring JNICALL
 Java_com_example_hellojni_HelloJni_stringFromJNI(JNIEnv* env,
                                                  jobject /* this */) {
diff --git a/hello-jni/app/src/main/java/com/example/hellojni/HelloJni.kt b/hello-jni/app/src/main/java/com/example/hellojni/HelloJni.kt
index 5454691..d8ca228 100644
--- a/hello-jni/app/src/main/java/com/example/hellojni/HelloJni.kt
+++ b/hello-jni/app/src/main/java/com/example/hellojni/HelloJni.kt
@@ -16,6 +16,7 @@
 package com.example.hellojni
 
 import android.os.Bundle
+import android.util.Log;
 import androidx.appcompat.app.AppCompatActivity
 import com.example.hellojni.databinding.ActivityHelloJniBinding
 
@@ -30,7 +31,10 @@ class HelloJni : AppCompatActivity() {
          */
         val binding = ActivityHelloJniBinding.inflate(layoutInflater)
         setContentView(binding.root)
-        binding.helloTextview.text = stringFromJNI()
+        // binding.helloTextview.text = stringFromJNI()
+        val t = java.lang.Thread(LoadHelloJniOnNewThread())
+        t.start()
+        t.join()
     }
 
     /*
@@ -61,8 +65,14 @@ class HelloJni : AppCompatActivity() {
      * at the installation time by the package manager.
      */
         init {
-            System.loadLibrary("hello-jni")
+            // System.loadLibrary("hello-jni")
         }
     }
 }
 
+class LoadHelloJniOnNewThread : Runnable {
+    override fun run() {
+        Log.w("*jonp*", "calling System.loadLibrary from new thread")
+        System.loadLibrary("hello-jni")
+    }
+}
\ No newline at end of file
```

Build, install, run:

```
% ./gradlew assembleDebug
% adb install app/build/outputs/apk/debug/app-debug.apk
% adb logcat > log.txt &
% adb shell am start com.example.hellojni/com.example.hellojni.HelloJni
```

Peruse `log.txt`.  For me, it contains the expected log messages:

```
W *jonp*  : calling System.loadLibrary from new thread
I *jonp*  : JNI_OnLoad called!
```

**Comment by jonpryor** [comment]
Current conjecture: https://discord.com/channels/732297728826277939/732297837953679412/1050805903554789467

When a [`java.lang.Thread` instance is created](https://android.googlesource.com/platform/libcore/+/refs/tags/android-platform-12.1.0_r10/ojluni/src/main/java/java/lang/Thread.java#686), it sets a thread-local `ClassLoader` instance to the `ClassLoader` of the "parent" (creating) thread.  This way, the `ClassLoader` from the UI thread "flows" to the new `Thread`.

That is *not* what happens in .NET Android & Xamarin.Android: here, Mono creates the thread, presumably through `posix_create()`, and…that's it.  No "flowing" of the `ClassLoader` is performed.  This results in/explains previous bugs/workarounds such as xamarin/java.interop@b4d44e4b3590a6a5a5e9bf4a0044eb2acf17bef6, wherein we need to *explicitly* pass around a `ClassLoader` so that new threads can actually find types located in `classes.dex` and *not* in `android.jar`.

(Which makes this a plausible 10+ year old design oversight.  Doh!)

Assuming this conjecture is correct, we need to "copy" `Thread.init2()` into `MonodroidRuntime::thread_start()` so that all MonoVM-created `Thread`s *also* inherit the `ClassLoader` of their calling thread: 
https://github.com/xamarin/xamarin-android/blob/2a16828a6c73d837e81c3a2d0c1ae2840d00873e/src/monodroid/jni/monodroid-glue.cc#L226-L242

**Comment by thenextman** [comment]
> @thenextman [wrote](https://github.com/xamarin/xamarin-android/issues/7616#issuecomment-1344475146):
> 
> > discovered that `LoadLibrary` fails in this manner if called from a background thread.
> 
> Could you please elaborate on what you mean by a "background thread"? Is this a `Java.Lang.Thread`-based background thread? A `System.Threading.Thread`-based background thread? A `Task`-based background thread? A POSIX `pthread_create()` based background thread? Something else?

@jonpryor In my case I just used `Thread.Start`, but the SO post I linked exhibits the same issue inside a C# `async` function.

If it's helpful at all I can upload my test project demonstrating this behaviour with Xamarin.

**Comment by jonpryor** [comment]
@thenextman: I don't think we'll need your test project anymore.  Confirmation that you're using `System.Threading.Thread` is sufficient.

We think we understand the problem, but it'll be some time before we can fix it.  (Holidays!)

In the meantime, calling `JavaSystem.LoadLibrary()` from the main/UI thread should allow your app to start.

Thanks!

**Comment by thenextman** [comment]
@jonathanpeppers I will try to re-evaluate this as time allows; for now the workaround of calling `JavaSystem.LoadLibrary` on the main thread at the appropriate time has served us quite well. For now I don't mind if the issue is closed; personally I don't have a hard need for this on .NET 9 as we'll be moving back to LTS at the end of the year. 

Once we're on .NET 10 I'll try to re-evaluate and confirm the fix. Maybe someone else needs this on .NET 9 and can confirm?
